org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/PasteEventHandler.java

```diff
@@ -334,6 +342,78 @@ private static String getEol(String text) {
 		return text.contains("\r\n") ? "\r\n" : "\n";
 	}
 
+	public static String handleFilePasteEvent(String path, String content, IProgressMonitor monitor) throws JavaModelException {
+		String desiredPath = "";
+		final ASTParser parser = ASTParser.newParser(AST.getJLSLatest());
+		parser.setResolveBindings(false);
+		parser.setKind(ASTParser.K_COMPILATION_UNIT);
+		parser.setStatementsRecovery(true);
+		parser.setBindingsRecovery(false);
+		parser.setIgnoreMethodBodies(true);
+		parser.setEnvironment(new String[0], new String[0], null, true);
+		Map<String, String> javaOptions = JavaCore.getOptions();
+		parser.setCompilerOptions(javaOptions);
+		parser.setSource(content.toCharArray());
+		CompilationUnit root = (CompilationUnit) parser.createAST(monitor);
+		// Check for package, import, and/or type declaration to determine if java file
+		if (root != null && (root.getPackage() != null || !root.imports().isEmpty() || !root.types().isEmpty())) {
+			String pathRoot = path;
+			if (root.getPackage() != null) {
+				String packageName = root.getPackage().getName().toString();
+				IJavaProject[] projects = ProjectUtils.getJavaProjects();
+				boolean pathFound = false;
+				String partialPathAddition = "";
+				for (int i = 0; i < projects.length; i++) {
+					if (!projects[i].getElementName().equals("jdt.ls-java-project") && !pathFound) {
+						IPackageFragment[] packages = projects[i].getPackageFragments();
+						for (int j = 0; j < packages.length; j++) {
+							if (packages[j].getKind() == IPackageFragmentRoot.K_SOURCE && !packages[j].isDefaultPackage()) {
+								pathRoot = packages[j].getParent().getResource().getLocation().toOSString();
+								String name = packages[j].getElementName();
+								if (name.equals(packageName)) {
+									desiredPath = packages[j].getResource().getLocation().toOSString();
+									pathFound = true;
+									break;
+								} else if (packageName.startsWith(name) && packageName.substring(0, name.length() + 1).endsWith(".")) {
+									String tempPath = packages[j].getResource().getLocation().toOSString();
+									if (tempPath.length() > desiredPath.length()) {
+										desiredPath = tempPath;
+										partialPathAddition = packageName.substring(name.length()).replaceAll("\\.", "/");
+									}
+								}
+							}
+						}
+					}
+				}
+
+				if (!pathFound && !partialPathAddition.isEmpty()) {
+					desiredPath = desiredPath + partialPathAddition;
+				} else if (!pathFound) {
+					packageName = packageName.replaceAll("\\.", "/");
```

Comment:
I would avoid using `/` as it's operating system dependent. Java has `File.separator` or `File.separatorChar` which should resolve correctly on each OS. You can also use `Paths.get(path1, path2, path3)` to join together various elements in a filesystem path.